music/src/gansynth/audio_utils.ts

```diff
@@ -0,0 +1,243 @@
+    reImBatch = tf.pad(reImBatch, [[0, 0], [0, 0], [1, 0], [0, 0]]);
+
+    // Interleave real and imaginary for javascript ISTFT.
+    // Hack to interleave [re0, im0, re1, im1, ...] with batchToSpace.
+    const crops = [[0, 0], [0, 0]];
+    const reImInterleave =
+        tf.batchToSpaceND(reImBatch, [1, 2], crops).reshape([128, 4096]);
+    // Convert Tensor to a Float32Array[]
+    return reImInterleave;
+  });
+  const reImArray = reImInterleave.dataSync();
+  const reIm = [] as Float32Array[];
+  for (let i = 0; i < 128; i++) {
+    reIm[i] = reImArray.slice(i * 4096, (i + 1) * 4096) as Float32Array;
+  }
+  return reIm;
+}
+
+async function reImToAudio(reIm: Float32Array[]) {
+  const ispecParams = {
+    nFFt: N_FFT,
+    winLength: N_FFT,
+    hopLength: N_HOP,
+    sampleRate: SAMPLE_RATE
+  };
+  return istft(reIm, ispecParams);
+}
+
+// TODO(jesseengel): Make it work for batch of spectrograms.
+export async function specgramsToAudio(specgrams: tf.Tensor4D) {
+  const reImArray = tf.tidy(() => {
+    // Synthesize audio
+    const magSlice = tf.slice(specgrams, [0, 0, 0, 0], [1, -1, -1, 1]).reshape([
+      1, 128, 1024
+    ]);
+    const magMel = magSlice as tf.Tensor3D;
+    const mag = melToLinear(magMel);
+
+    const ifreqSlice = tf.slice(specgrams, [0, 0, 0, 1], [
+                           1, -1, -1, 1
+                         ]).reshape([1, 128, 1024]);
+    const ifreq = ifreqSlice as tf.Tensor3D;
+    const phase = ifreqToPhase(ifreq);
+
+    // Reflect all frequencies except for the Nyquist, which is shared between
+    // positive and negative frequencies for even nFft.
+    let real = mag.mul(tf.cos(phase));
+    const mirrorReal = tf.reverse(real.slice([0, 0, 0], [1, 128, 1023]), 2);
+    real = tf.concat([real, mirrorReal], 2);
+
+    // Reflect all frequencies except for the Nyquist, take complex conjugate of
+    // the negative frequencies.
+    let imag = mag.mul(tf.sin(phase));
+    const mirrorImag = tf.reverse(imag.slice([0, 0, 0], [1, 128, 1023]), 2);
+    imag = tf.concat([imag, tf.mul(mirrorImag, -1.0)], 2);
+    return [real, imag];
+  });
+  const reIm = await interleaveReIm(reImArray[0], reImArray[1]);
+  const audio = await reImToAudio(reIm);
+  return audio;
+}
+
+//------------------------------------------------------------------------------
+// FFT Code
+//------------------------------------------------------------------------------
+export function fft(y: Float32Array) {
+  const fft = new FFT(y.length);
+  const out = fft.createComplexArray();
+  const data = fft.toComplexArray(y);
+  fft.transform(out, data);
+  return out;
+}
+
+export function stft(y: Float32Array, params: SpecParams): Float32Array[] {
+  const nFft = params.nFft || 2048;
+  const winLength = params.winLength || nFft;
+  const hopLength = params.hopLength || Math.floor(winLength / 4);
+
+  const fftWindow = hannWindow(winLength);
+
+  // Pad the window to be the size of nFft.
```

Comment:
remove here and below?

Reply:
I just removed the whole fft because I don't actually use it.